README.md
[![Build Status](https://travis-ci.org/express42-cookbooks/openvpn.svg?branch=master)](https://travis-ci.org/express42-cookbooks/openvpn)

# Description

Installs and configures OpenVPN.

# Requirements

## Platform:

* Ubuntu
* RHEL
* CentOS

## Gems:

* knife-openvpn

# Attributes

* `node['openvpn']['server_name']` -  Defaults to `"default"`.
* `node['openvpn']['install_epel']` -  Defaults to `true`.
* `node['openvpn']['ip_forward']` -  Defaults to `true`.
* `node['openvpn']['iptables']['postrouting']` -  Defaults to `true` for `RHEL` based platforms.
* `node['openvpn']['iptables']['interface']` -  Defaults to `eth0`.
* `node['openvpn']['default']['remote_host']` -  Defaults to `"vpn.example.com"`.
* `node['openvpn']['default']['server_ip']` -  Defaults to `"127.0.0.1"`.
* `node['openvpn']['default']['port']` -  Defaults to `"1194"`.
* `node['openvpn']['default']['proto']` -  Defaults to `"udp"`.
* `node['openvpn']['default']['dev']` -  Defaults to `"tun"`.
* `node['openvpn']['default']['mode']` -  Defaults to `"routed"`.
* `node['openvpn']['default']['netmask']` -  Defaults to `"255.255.255.0"`.
* `node['openvpn']['default']['subnet']` -  Defaults to `"127.0.1.0"`.
* `node['openvpn']['default']['network_bridge']` -  Defaults to `"br0"`.
* `node['openvpn']['default']['network_interface']` -  Defaults to `"eth0"`.
* `node['openvpn']['default']['dhcp_start']` -  Defaults to `"127.0.0.100"`.
* `node['openvpn']['default']['dhcp_end']` -  Defaults to `"127.0.0.150"`.
* `node['openvpn']['default']['verb']` -  Defaults to `"3"`.
* `node['openvpn']['default']['push']` -  Defaults to `"[ ... ]"`.
* `node['openvpn']['default']['duplicate_cn']` -  Defaults to `"false"`.
* `node['openvpn']['default']['client_to_client']` -  Defaults to `"false"`.
* `node['openvpn']['default']['keepalive_interval']` -  Defaults to `"10"`.
* `node['openvpn']['default']['keepalive_timeout']` -  Defaults to `"60"`.
* `node['openvpn']['default']['comp_lzo']` -  Defaults to `"true"`.
* `node['openvpn']['default']['link_mtu']` -  Defaults to `"nil"`.
* `node['openvpn']['default']['tun_mtu']` -  Defaults to `"nil"`.
* `node['openvpn']['default']['cipher']` -  Defaults to `"false"`.
* `node['openvpn']['default']['redirect_gateway']` -  Defaults to `"false"`.
* `node['openvpn']['default']['push_dns_server']` -  Defaults to `"false"`.
* `node['openvpn']['default']['script_security']` -  Defaults to `"1"`.
* `node['openvpn']['default']['use_tls_auth']` -  Defaults to `"true"`.
* `node['openvpn']['default']['chroot']` -  Defaults to `"false"`.
* `node['openvpn']['default']['client_config_dir']` -  Defaults to `"false"`.
* `node['openvpn']['default']['ccd_exclusive']` -  Defaults to `"false"`.
* `node['openvpn']['default']['users']` -  Defaults to `"[ ... ]"`.
* `node['openvpn']['default']['revoked_users']` -  Defaults to `"[ ... ]"`.
* `node['openvpn']['default']['ifconfig_pool_persist']` - Defaults to `"true"`

# Recipes

* openvpn::default - Installs and configures OpenVPN.

# Server Modes

* Routed

For routed network you must define vpn ```subnet```, like in previous example

* Bridged

Bridged setup need more configuration and configured network bridge on your server

```
"default_attributes": {
  "openvpn": {
    "server_name": "office",
    "office": {
      "remote_host": "vpn.example.com",
      "server_ip": "10.90.5.5",
      "port": "443",
      "proto": "tcp",
      "dev": "tap",
      "verb": "3",
      "mode": "bridged",
      "script_security": "2",
      "dhcp_start": "10.90.5.100",
      "dhcp_end": "10.90.5.240",
      "network_bridge": "br0",
      "network_interface": "eth0"
    }
  }
}

```

See fixture cookbook in `tests/fixtures/cookbooks`.


# To-do

1. Revoke access
2. Import existing certs/keys


# Usage

For example you want to setup vpn server and call it ```office```

* Ensure that you have ```.chef/encrypted_data_bag_secret```.
Otherwise you can generate one with ```openssl rand -base64 512 > .chef/encrypted_data_bag_secret```

* Install knife plugin:

  ```
  gem install knife-openvpn
  ```

* Create server certificate authority, server cert/key, DH params:

  ```
  knife openvpn server create office
  ```

  ```office``` - is a name of vpn-server, there is some limitations on this: no dots, no commas, no spaces, no special symbols for reasons.

* Great, now check ```data_bags``` directory, you will find new databag ```openvpn-office``` with few items for ca, dh, cert/key pair and some openssl config. Now it is time to upload it to Chef server:

  ```
  knife data bag create openvpn-office --secret-file=.chef/encrypted_data_bag_secret
  knife data bag from file openvpn-office data_bags/openvpn-office/*
  ```

* Add ```recipe[openvpn]``` to node run_list, and override default attributes:

  ```
  "run_list": [
  "recipe[openvpn]"
  ],
  "default_attributes": {
    "openvpn": {
      "server_name": "office",
      "office": {
        "remote_host": "vpn.example.com",
        "server_ip": "10.90.5.5",
        "subnet": "10.200.1.0",
        "port": "443",
        "proto": "tcp",
        "dev": "tun",
        "verb": "3",
        "push": [
        "route 10.90.0.0 255.255.255.0",
        "route 10.90.1.0 255.255.255.0"
        ]
      }
    }
  }

  ```

  Chef, run!

* When server is up and running we can add some users to start use it.
No moar certificate management pain:

  ```
  knife openvpn user create office john
  knife data bag from file openvpn-office data_bags/openvpn-office/john.json
  ```

* Export vpn-client data and send it to John:

  ```
  knife openvpn user export office john
  ```
resulting archive contains config (.ovpn), ca cert, John's cert and key

* Revokation of user certificate is also possible:
  ```
  knife openvpn user revoke office john
  knife data bag from file openvpn-office data_bags/openvpn-office/openvpn-crl.json
  ```


# License and Maintainer

Maintainer:: LLC Express 42 (<cookbooks@express42.com>)

License:: MIT


CHANGELOG.md
# CHANGELOG

This file is used to list changes made in each version of the openvpn cookbook.

0.1.6
-----
- New attribute ifconfig_pool_persist (bool). 

0.1.5
-----
- Recreate pem files if they where changed
- Restart openvpn service if server files were changed


0.1.4
-----
- Setup iptables and sysctl
- Initial RHEL support

0.1.3
-----
- Remove knife plugin

0.1.2
-----
- Fix Rubocop warnings
- Add Test Kitchen, Rubocop, Berkshelf, Serverspec and Travis CI
- Add encrypted data bags to integration test
- Update docs

0.1.1
-----
- Fix permission for crl file

0.1.0
-----
- Initial release of openvpn

- - -
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


test/fixtures/cookbooks/openvpn_test/metadata.rb
name 'openvpn_test'

depends 'apt'
depends 'yum-epel'
depends 'openvpn'


test/integration/routed/serverspec/openvpn_spec.rb
require 'spec_helper'

describe package('openvpn') do
  it { should be_installed }
end

describe user('openvpn') do
  it { should exist }
  it { should belong_to_group 'openvpn' }
end

describe file('/etc/openvpn/default') do
  it { should be_directory }
end

describe file('/etc/openvpn/default/keys') do
  it { should be_directory }
end

describe file('/var/log/openvpn') do
  it { should be_directory }
end

describe file('/etc/openvpn/default/keys/ca.crt') do
  it { should be_file }
end

describe file('/etc/openvpn/default/keys/dh.pem') do
  it { should be_file }
end

describe file('/etc/openvpn/default/keys/crl.pem') do
  it { should be_file }
end

describe file('/etc/openvpn/default/keys/server.crt') do
  it { should be_file }
end

describe file('/etc/openvpn/default/keys/server.key') do
  it { should be_file }
end

describe file('/etc/openvpn/default.conf') do
  it { should be_file }
  it { should contain 'Generated by Chef' }
  it { should contain 'port 1194' }
  it { should contain 'proto udp' }
  it { should contain 'dev tun' }
  it { should contain 'server 127.0.1.0 255.255.255.0' }
  it { should contain 'local 127.0.0.1' }
end

describe service('openvpn') do
  it { should be_enabled }
  it { should be_running }
end

describe port(1194) do
  it { should be_listening }
end


test/integration/bridged/serverspec/openvpn_spec.rb
require 'spec_helper'

describe interface('br0') do
  it { should exist }
end

describe package('openvpn') do
  it { should be_installed }
end

describe package('bridge-utils') do
  it { should be_installed }
end

describe user('openvpn') do
  it { should exist }
  it { should belong_to_group 'openvpn' }
end

describe file('/etc/openvpn/default') do
  it { should be_directory }
end

describe file('/etc/openvpn/default/keys') do
  it { should be_directory }
end

describe file('/var/log/openvpn') do
  it { should be_directory }
end

describe file('/etc/openvpn/default/keys/ca.crt') do
  it { should be_file }
end

describe file('/etc/openvpn/default/keys/dh.pem') do
  it { should be_file }
end

describe file('/etc/openvpn/default/keys/crl.pem') do
  it { should be_file }
end

describe file('/etc/openvpn/default/keys/server.crt') do
  it { should be_file }
end

describe file('/etc/openvpn/default/keys/server.key') do
  it { should be_file }
end

describe file('/etc/openvpn/default.conf') do
  it { should be_file }
  it { should contain 'Generated by Chef' }
  it { should contain 'port 1194' }
  it { should contain 'proto udp' }
  it { should contain 'dev tap' }
  it { should contain 'server-bridge 127.0.0.1 255.255.255.0 127.0.0.100 127.0.0.150' }
  it { should contain 'up "/etc/openvpn/up.sh br0 eth0"' }
  it { should contain 'down "/etc/openvpn/down.sh br0 eth0"' }
end

describe file('/etc/openvpn/up.sh') do
  it { should be_file }
end

describe file('/etc/openvpn/down.sh') do
  it { should be_file }
end

describe service('openvpn') do
  it { should be_enabled }
  it { should be_running }
end

describe port(1194) do
  it { should be_listening }
end


knife_plugin.rb
#
# Cookbook Name:: openvpn
# OpenVPN knife plugin
#
# Copyright 2013, LLC Express 42
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
#
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#

module OpenvpnPlugin
  class Openvpn < Chef::Knife
    def run
      ui.info 'knife openvpn (user|server) action ARGS OPTS'
    end

    deps do
      require 'chef/encrypted_data_bag_item'
      require 'json'
      require 'openssl'
    end

    def check_databag_secret
      databag_secret_file = File.join(Dir.pwd, '.chef/encrypted_data_bag_secret')
      unless File.exist? databag_secret_file
        fail_with "Can't find encrypted databag secret file at #{databag_secret_file}."
      end
    end

    def check_existing_databag(server_name, fail_if_exists = false)
      databag_directory = File.join(Dir.pwd, "data_bags/openvpn-#{server_name}")
      if File.directory? databag_directory
        if fail_if_exists # databag exists and we want to create new
          fail_with "Data bag directory #{databag_directory} already exists."
        end
      else
        unless fail_if_exists # no such databag, but we want to use it
          fail_with "Data bag #{databag_directory} not exists."
        end
      end
    end

    def fail_with(error_message)
      ui.error "Error: #{error_message}"
      exit 1
    end

    def make_name(cn, cert_config)
      name = OpenSSL::X509::Name.new
      name.add_entry 'CN', cn
      %w(C L O OU ST mail).each { |entry| name.add_entry(entry, cert_config[entry]) }
      name
    end

    def get_databag_secret
      databag_secret_file = File.join(Dir.pwd, '.chef/encrypted_data_bag_secret')
      secret = Chef::EncryptedDataBagItem.load_secret(databag_secret_file)
      secret
    end

    def get_extensions_factory(subject_cert, issuer_cert)
      factory = OpenSSL::X509::ExtensionFactory.new
      factory.subject_certificate = subject_cert
      factory.issuer_certificate = issuer_cert
      factory
    end

    def add_ca_extensions(ca_cert)
      ef = get_extensions_factory ca_cert, ca_cert
      ca_cert.add_extension(ef.create_extension('basicConstraints', 'CA:TRUE', true))
      ca_cert.add_extension(ef.create_extension('keyUsage', 'keyCertSign, cRLSign', true))
      ca_cert.add_extension(ef.create_extension('subjectKeyIdentifier', 'hash', false))
      ca_cert.add_extension(ef.create_extension('authorityKeyIdentifier', 'keyid:always', false))
    end

    def add_endentity_extensions(entity_cert, ca_cert)
      ef = get_extensions_factory entity_cert, ca_cert
      entity_cert.add_extension(ef.create_extension('keyUsage', 'digitalSignature', true))
      entity_cert.add_extension(ef.create_extension('subjectKeyIdentifier', 'hash', false))
    end

    def generate_cert_and_key(subject, cert_config, selfsigned = false, ca_cert = nil, ca_key = nil)
      key = OpenSSL::PKey::RSA.generate(cert_config['rsa_keysize'])
      cert = OpenSSL::X509::Certificate.new
      cert.version = 2
      cert.serial = Time.now.to_i
      cert.public_key = key.public_key

      cert.not_after = Time.now + (cert_config['years_to_expire'] * 365 * 24 * 60 * 60)
      cert.not_before = Time.now - (24 * 60 * 60)

      if selfsigned
        cert.subject = subject
        cert.issuer = subject
        add_ca_extensions(cert)
        cert.sign(key, OpenSSL::Digest::SHA1.new)
      else
        if ca_cert.nil? || ca_key.nil?
          fail_with "CA key or cert isn't specified"
        end
        cert.subject = subject
        cert.issuer = ca_cert.subject
        add_endentity_extensions(cert, ca_cert)
        cert.sign(ca_key, OpenSSL::Digest::SHA1.new)
      end

      [cert, key]
    end

    def issue_crl(revoke_info, serial, lastup, nextup, extensions,
                 issuer, issuer_key, digest)
      crl = OpenSSL::X509::CRL.new
      crl.issuer = issuer.subject
      crl.version = 1
      crl.last_update = lastup
      crl.next_update = nextup
      revoke_info.each do|rserial, time, reason_code|
        revoked = OpenSSL::X509::Revoked.new
        revoked.serial = rserial
        revoked.time = time
        enum = OpenSSL::ASN1::Enumerated(reason_code)
        ext = OpenSSL::X509::Extension.new('CRLReason', enum)
        revoked.add_extension(ext)
        crl.add_revoked(revoked)
      end
      ef = OpenSSL::X509::ExtensionFactory.new
      ef.issuer_certificate = issuer
      ef.crl = crl
      crlnum = OpenSSL::ASN1::Integer(serial)
      crl.add_extension(OpenSSL::X509::Extension.new('crlNumber', crlnum))
      extensions.each do|oid, value, critical|
        crl.add_extension(ef.create_extension(oid, value, critical))
      end
      crl.sign(issuer_key, digest)
      crl
   end

    def load_cert_and_key(cert_str, key_str)
      cert = OpenSSL::X509::Certificate.new cert_str
      key = OpenSSL::PKey::RSA.new key_str
      [cert, key]
    end

    def get_databag_path(server_name)
      directory_path = File.join(Dir.pwd, "data_bags/openvpn-#{server_name}")
      directory_path
    end

    def get_databag_name(server_name)
      databag_name = "openvpn-#{server_name}"
      databag_name
    end

    def save_databag_item(id, server_name, item_hash)
      databag_path = get_databag_path server_name
      item_hash['id'] = id
      item_path = File.join(databag_path, "#{id}.json")
      secret = get_databag_secret
      encrypted_data = Chef::EncryptedDataBagItem.encrypt_data_bag_item(item_hash, secret)
      unless File.exist? item_path
        File.write item_path, JSON.pretty_generate(encrypted_data)
      else
        fail_with "#{item_path} already exists"
      end
    end

    def load_databag_item(databag_name, item_id)
      secret = get_databag_secret
      # puts "Loading [#{databag_name}:#{item_id}]"
      item = Chef::EncryptedDataBagItem.load(databag_name, item_id, secret)
      item
    end
  end

  class OpenvpnServerCreate < Openvpn
    banner 'knife openvpn server create NAME (options)'
    deps do
      require 'readline'
    end

    def run
      check_arguments
      vpn_server_name = name_args.first
      check_existing_databag vpn_server_name, true
      check_databag_secret
      create_new_server vpn_server_name
    end

    def create_new_server(vpn_server_name)
      now = Time.at(Time.now.to_i)
      cert_config = ask_for_cert_config
      ca_subject = make_name 'CA', cert_config
      ca_cert, ca_key = generate_cert_and_key ca_subject, cert_config, true
      server_subject = make_name vpn_server_name, cert_config
      server_cert, server_key = generate_cert_and_key server_subject, cert_config, false, ca_cert, ca_key
      dh_params = make_dh_params cert_config
      crl = issue_crl([], 1, now, now + 3600, [], ca_cert, ca_key, OpenSSL::Digest::SHA1.new)
      databag_path = get_databag_path vpn_server_name
      ui.info "Creating data bag directory at #{databag_path}"
      create_databag_dir vpn_server_name
      save_databag_item('openvpn-config', vpn_server_name, cert_config)
      save_databag_item('openvpn-ca', vpn_server_name, 'cert' => ca_cert.to_pem, 'key' => ca_key.to_pem)
      save_databag_item('openvpn-crl', vpn_server_name, 'crl' => crl.to_pem, 'revoke_info' => [])

      save_databag_item('openvpn-server', vpn_server_name, 'cert' => server_cert.to_pem, 'key' => server_key.to_pem)
      save_databag_item('openvpn-dh', vpn_server_name, 'dh' => dh_params.to_pem)
    end

    def check_arguments
      unless name_args.size == 1
        fail_with 'Specify NAME of new openvpn server!'
      end
    end

    def create_databag_dir(server_name)
      databag_path = get_databag_path server_name
      Dir.mkdir(databag_path, 0755)
      databag_path
    end

    def read_with_prompt_and_default(prompt, default)
      answer = Readline.readline("#{prompt} [#{default}]: ").strip
      if answer.empty?
        default
      else
        answer
      end
    end

    def make_dh_params(cert_config)
      keysize = cert_config['dh_keysize']
      dh_params = OpenSSL::PKey::DH.new keysize
      dh_params
    end

    def ask_for_cert_config
      cert_config = {}
      strings_prompt_default = [
        ['C', 'Country Name', 'RU'],
        ['ST', 'State or Province Name', 'MSK'],
        ['L', 'Locality Name', 'Moscow'],
        ['O', 'Organization Name', 'Express 42'],
        ['OU', 'Organizational Unit Name', 'OPS'],
        ['mail', 'Email', 'ops@example.com']
      ]
      numeric_prompt_default = [
        ['rsa_keysize', 'RSA key size (1024/2048/4096)', '2048'],
        ['dh_keysize', 'DH key size (1024/2048/4096)', '1024'],
        ['years_to_expire', 'Expiration (in years from now)', '5']
      ]
      strings_prompt_default.each { |entry| cert_config[entry[0]] = read_with_prompt_and_default(entry[1], entry[2]) }
      numeric_prompt_default.each { |entry| cert_config[entry[0]] = read_with_prompt_and_default(entry[1], entry[2]).to_i }
      %w(rsa_keysize dh_keysize).each do |keysize|
        unless [1024, 2048, 4096].include? cert_config[keysize]
          fail_with "Wrong value for #{keysize}, must be one of 1024/2048/4096"
        end
      end
      cert_config
    end
  end

  class OpenvpnUserCreate < Openvpn
    banner 'knife openvpn user create SERVERNAME USERNAME (options)'

    def run
      check_arguments
      server_name = name_args[0]
      user_name = name_args[1]
      check_existing_databag server_name, false
      check_databag_secret
      create_new_user server_name, user_name
    end

    def create_new_user(server_name, user_name)
      databag_name = get_databag_name server_name
      ca_item = load_databag_item(databag_name, 'openvpn-ca')
      ca_cert, ca_key = load_cert_and_key ca_item['cert'], ca_item['key']
      config_item = load_databag_item(databag_name, 'openvpn-config')
      cert_config = config_item.to_hash
      user_subject = make_name user_name, cert_config
      user_cert, user_key = generate_cert_and_key user_subject, cert_config, false, ca_cert, ca_key
      save_databag_item(user_name, server_name, 'cert' => user_cert.to_pem, 'key' => user_key.to_pem)
      ui.info "Done, now you can upload #{databag_name}/#{user_name}.json"
    end

    def check_arguments
      unless name_args.size == 2
        fail_with 'Specify SERVERNAME and USERNAME for new openvpn user!'
      end
    end
  end

  class OpenvpnUserExport < Openvpn
    banner 'knife openvpn user export SERVERNAME USERNAME (options)'

    deps do
      require 'chef/search/query'
    end

    def run
      check_arguments
      server_name = name_args[0]
      user_name = name_args[1]
      check_existing_databag server_name, false
      check_databag_secret
      export_user server_name, user_name
    end

    def export_user(server_name, user_name)
      databag_name = get_databag_name server_name
      ca_item = load_databag_item(databag_name, 'openvpn-ca')
      ca_cert, ca_key = load_cert_and_key ca_item['cert'], ca_item['key']

      user_item = load_databag_item(databag_name, user_name)
      user_cert, user_key = load_cert_and_key user_item['cert'], user_item['key']
      tmpdir = Dir.mktmpdir
      ui.info "tmpdir: #{tmpdir}"
      begin
        user_dir = "#{tmpdir}/#{user_name}-vpn"
        Dir.mkdir user_dir
        ui.info "userdir: #{user_dir}"
        export_file "#{user_dir}/ca.crt", ca_cert.to_pem
        export_file "#{user_dir}/#{user_name}.crt", user_cert.to_pem
        export_file "#{user_dir}/#{user_name}.key", user_key.to_pem
        config_content = generate_client_config server_name, user_name
        export_file "#{user_dir}/#{user_name}.ovpn", config_content
        exitcode = system("cd #{tmpdir} && tar cfz /tmp/#{user_name}-vpn.tar.gz *")
        if exitcode
          ui.info "Done, archive at /tmp/#{user_name}-vpn.tar.gz"
        else
          ui.error "Something went wrong, cant create archive at /tmp/#{user_name}-vpn.tar.gz"
        end
      ensure
        FileUtils.rm_rf(tmpdir)
      end
    end

    def export_file(file_path, content)
      File.write file_path, content
      FileUtils.chmod 'u=wr,go-wr', file_path
    end

    def generate_client_config(server_name, user_name)
      query = "openvpn_server_name:#{server_name}"
      query_nodes = Chef::Search::Query.new
      search_result = query_nodes.search('node', query)[0]
      unless search_result.length == 1
        fail_with "Found #{search_result.length} vpn servers for #{server_name}"
      end
      config_content = ''
      newline = "\n"
      node = search_result[0]
      config_content << 'client' << newline
      config_content << "dev  #{node['openvpn'][server_name]['dev']}" << newline
      config_content << "proto  #{node['openvpn'][server_name]['proto']}" << newline
      config_content << "remote  #{node['openvpn'][server_name]['remote_host']} "
      config_content << "#{node['openvpn'][server_name]['port']}" << newline
      config_content << "verb  #{node['openvpn'][server_name]['verb']}" << newline
      config_content << 'comp-lzo' << newline
      config_content << 'ca ca.crt' << newline
      config_content << "cert #{user_name}.crt" << newline
      config_content << "key #{user_name}.key" << newline
      config_content << 'nobind' << newline
      config_content << 'persist-key' << newline
      config_content << 'persist-tun' << newline
      config_content
    end

    def check_arguments
      unless name_args.size == 2
        fail_with 'Specify SERVERNAME and USERNAME for new openvpn user!'
      end
    end
  end

  class OpenvpnUserRevoke < Openvpn
    banner 'knife openvpn user revoke SERVERNAME USERNAME'

    deps do
      require 'chef/search/query'
    end

    def run
      check_arguments
      server_name = name_args[0]
      user_name = name_args[1]
      check_existing_databag server_name, false
      check_databag_secret
      revoke_user server_name, user_name
    end

    def revoke_user(server_name, user_name)
      now = Time.at(Time.now.to_i)
      databag_name = get_databag_name server_name
      ca_item = load_databag_item(databag_name, 'openvpn-ca')
      ca_cert, ca_key = load_cert_and_key ca_item['cert'], ca_item['key']
      begin
        crl_item = load_databag_item(databag_name, 'openvpn-crl')
        old_crl = OpenSSL::X509::CRL.new crl_item['crl']
        revoke_info = crl_item['revoke_info']
      rescue
        old_crl = issue_crl([], 1, now, now + 3600, [], ca_cert, ca_key, OpenSSL::Digest::SHA1.new)
        revoke_info = []
      end
      user_item = load_databag_item(databag_name, user_name)
      user_cert, user_key = load_cert_and_key user_item['cert'], user_item['key']
      user_revoke_info = [[user_cert.serial, now, 0]]
      new_revoke_info = revoke_info + user_revoke_info
      new_crl = add_user_to_crl ca_cert, ca_key, old_crl, new_revoke_info
      save_databag_item('openvpn-crl', server_name, 'crl' => new_crl.to_pem, 'revoke_info' => new_revoke_info)
      ui.info "revoked #{user_name}, do not forget to upload CRL databag item"
    end

    def add_user_to_crl(ca_cert, ca_key, old_crl, revoke_info)
      new_crl = issue_crl(revoke_info, old_crl.version + 1, Time.at(Time.now.to_i), Time.at(Time.now.to_i) + 3600, [], ca_cert, ca_key, OpenSSL::Digest::SHA1.new)
      new_crl
    end

    def check_arguments
      unless name_args.size == 2
        fail_with 'Specify SERVERNAME and USERNAME for existing openvpn user!'
      end
    end
  end
end


attributes/default.rb
default['openvpn']['server_name'] = 'default'
default['openvpn']['default']['users'] = []
default['openvpn']['default']['revoked_users'] = []
default['openvpn']['default']['remote_host'] = 'vpn.example.com'

# Network
default['openvpn']['default']['server_ip'] = '127.0.0.1'
default['openvpn']['default']['port'] = '1194'
default['openvpn']['default']['proto'] = 'udp'
default['openvpn']['default']['dev'] = 'tun'
default['openvpn']['default']['mode'] = 'routed'
default['openvpn']['default']['netmask'] = '255.255.255.0'
default['openvpn']['default']['subnet'] = '127.0.0.0'
default['openvpn']['default']['network_bridge'] = 'br0'
default['openvpn']['default']['network_interface'] = 'eth0'
default['openvpn']['default']['dhcp_start'] = '127.0.0.100'
default['openvpn']['default']['dhcp_end'] = '127.0.0.150'
default['openvpn']['default']['verb'] = '3'


default['openvpn']['default']['push'] = []
default['openvpn']['default']['duplicate_cn'] = false
default['openvpn']['default']['client_to_client'] = false 
default['openvpn']['default']['keepalive_interval'] = 10
default['openvpn']['default']['keepalive_timeout'] = 60
default['openvpn']['default']['comp_lzo'] = true
default['openvpn']['default']['link_mtu'] = nil
default['openvpn']['default']['tun_mtu'] = nil
default['openvpn']['default']['cipher'] = false
default['openvpn']['default']['redirect_gateway'] = false
default['openvpn']['default']['push_dns_server'] = false
default['openvpn']['default']['script_security'] = 1
default['openvpn']['default']['ifconfig_pool_persist'] = true